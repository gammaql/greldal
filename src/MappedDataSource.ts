import { getTypeAccessorError } from "./errors";
import { TypeGuard, Dict, NNil, Maybe, MaybeArrayItem, ReturnType } from "./util-types";
import { isString, transform, camelCase, upperFirst, snakeCase, forEach, reduce } from "lodash";
import * as t from "io-ts";
import * as Knex from "knex";
import _debug from "debug";
import { GraphQLInputType, GraphQLOutputType } from "graphql";
import { MappedField } from "./MappedField";
import { FieldMapping } from "./FieldMapping";
import { MappedAssociation } from "./MappedAssociation";
import { singularize, pluralize } from "inflection";
import {
    deriveDefaultShallowOutputType,
    deriveDefaultOutputType,
    deriveDefaultShallowInputType,
    derivePaginatedOutputType,
} from "./graphql-type-mapper";
import { assertSupportedConnector, globalConnector, assertConnectorConfigured } from "./connector";
import { MemoizeGetter } from "./utils";
import { StoreQueryParams } from "./SingleSourceQueryOperationResolver";
import { ReverseMapper } from "./ReverseMapper";
import { AliasHierarchyVisitor } from "./AliasHierarchyVisitor";
import { DataSourceMapping } from "./DataSourceMapping";

const debug = _debug("greldal:MappedDataSource");

type AssociationsIn<T extends DataSourceMapping> = ReturnType<NNil<T["associations"]>>;

type FieldsIn<T extends DataSourceMapping> = ReturnType<NNil<T["fields"]>>;

type AssociationKeysIn<T extends DataSourceMapping> = keyof AssociationsIn<T>;

type FieldKeysIn<T extends DataSourceMapping> = keyof FieldsIn<T>;

type AssociatedDataSource<T extends DataSourceMapping, K extends AssociationKeysIn<T>> = AssociationsIn<T>[K]["target"];

type ShallowEntityType<T extends DataSourceMapping> = {
    [K in FieldKeysIn<T>]: t.TypeOf<FieldsIn<T>[K]["type"]>
};

type NestedEntityType<T extends DataSourceMapping> = ShallowEntityType<T> &
    { [K in AssociationKeysIn<T>]: AssociatedDataSource<T, K>["EntityType"] };

/**
 * Represents mapping between a relational data source and associated GraphQL types
 * originating from the schema of this data source.
 *
 * @api-category MapperClass
 */
export class MappedDataSource<T extends DataSourceMapping = any> {
    fields: Dict<MappedField>;
    associations: Dict<MappedAssociation>;

    constructor(private mapping: T) {
        this.fields = mapping.fields ? mapping.fields(this) : {};
        this.associations = mapping.associations ? mapping.associations(this) : {};
        debug(
            "Mapped data source with name: %s, fields: %O, associations: %O",
            this.mappedName,
            Object.keys(this.fields),
            Object.keys(this.associations),
        );
        if (mapping.connector) {
            assertSupportedConnector(mapping.connector);
        }
    }

    /**
     * Knex instance used to connect to data source.
     *
     * This can be a data-source specific connector (if provided in the mapping configuration), and if not,
     * will fall back to global connector.
     *
     * Throws if a connector is not found.
     */
    get connector(): Knex {
        return assertConnectorConfigured(this.mapping.connector || globalConnector);
    }

    /**
     * Get a knex query builder for this data source
     *
     * This internally aliases tables for ease of reverse-mapping.
     */
    rootQueryBuilder(aliasHierarchyVisitor?: Maybe<AliasHierarchyVisitor>): Knex.QueryBuilder {
        if (this.mapping.rootQuery) return this.mapping.rootQuery(aliasHierarchyVisitor);
        return aliasHierarchyVisitor
            ? this.connector(`${this.storedName} as ${aliasHierarchyVisitor.alias}`)
            : this.connector(this.storedName);
    }

    /**
     * Get list of fields representing columns covered by primary key constraint
     */
    get primaryFields() {
        return Object.values<MappedField<any>>(this.fields).filter(f => f.isPrimary);
    }

    /**
     * Name of the GraphQL output type representing an entity from this data source. Also used in other GraphQL output
     * types for this data source.
     */
    @MemoizeGetter
    get mappedName() {
        return (isString as TypeGuard<string>)(this.mapping.name)
            ? upperFirst(camelCase(singularize(this.mapping.name)))
            : this.mapping.name.mapped;
    }

    /**
     * Name of the GraphQL output type representing a shallow entity (containing only fields and not associations) from this data source.
     */
    @MemoizeGetter
    get shallowMappedName() {
        return `Shallow${this.mappedName}`;
    }

    /**
     * Name of the table/view backing this data source
     */
    @MemoizeGetter
    get storedName() {
        return (isString as TypeGuard<string>)(this.mapping.name)
            ? snakeCase(pluralize(this.mapping.name))
            : this.mapping.name.stored;
    }

    /**
     * Name of the GraphQL output type representing a page container that wraps page specific metadata along with
     * entities in the page (in a paginated query)
     */
    @MemoizeGetter
    get pageContainerName() {
        return `${this.mappedName}PageContainer`;
    }

    /**
     * Name of the GraphQL output type representing a page container that wraps page specific metadata along with
     * shallow entities in the page (in a paginated query)
     */
    @MemoizeGetter
    get shallowPageContainerName() {
        return `${this.shallowMappedName}PageContainer`;
    }

    /**
     * Name of the GraphQL output type representing a page that wraps a subset of result entities (in a paginated query)
     */
    @MemoizeGetter
    get pageName() {
        return `${this.mappedName}Page`;
    }

    /**
     * Name of the GraphQL output type representing a page that wraps a subset of shallow result entities (in a paginated query)
     */
    @MemoizeGetter
    get shallowPageName() {
        return `${this.shallowMappedName}Page`;
    }

    /**
     * List of names of the columns in the data source which are mapped through the fields in the data source mapping
     *
     * It is not necessary that all columns of backing table are covered by fields of the data source.
     */
    @MemoizeGetter
    get storedColumnNames(): string[] {
        return transform(
            this.fields,
            (result: string[], field: MappedField<MappedDataSource<T>, FieldMapping<any, any>>) => {
                if (field.isMappedFromColumn) {
                    result.push(field.sourceColumn!);
                }
            },
            [],
        );
    }

    /**
     * io-ts type props for field properties of a member entity
     */
    @MemoizeGetter
    get shallowEntityProps(): t.Props & Dict<t.Type<any>> {
        return transform(this.fields, (result, field, name) => {
            result[name] = field.type;
            return result;
        });
    }

    /**
     * io-ts type props for association properties of a member entity
     */
    @MemoizeGetter
    get associationProps(): t.Props & Dict<t.Type<any>> {
        const result: t.Props & Dict<t.Type<any>> = {};
        forEach(this.associations, (association, name) => {
            result[name] = association.target.entityType;
        });
        return transform(
            this.associations,
            (result: t.Props, association: MappedAssociation<MappedDataSource<T>>, name: string) => {
                result[name] = association.target.entityType;
            },
            {},
        );
    }

    /**
     * io-ts type props for all properties (from fields and associations)
     */
    @MemoizeGetter
    get entityProps(): t.Props & Dict<t.Type<any>> {
        return {
            ...this.shallowEntityProps,
            ...this.associationProps,
        };
    }

    /**
     * io-ts runtime type for shallow member entity (excludes associations) from this data source.
     */
    @MemoizeGetter
    get shallowEntityType() {
        return t.type(this.shallowEntityProps);
    }

    /**
     * io-ts runtime type for member entity from this data source.
     */
    @MemoizeGetter
    get entityType() {
        return t.type(this.entityProps);
    }

    /**
     * Getter to extract static type of Shallow member entity
     *
     * This is only useful in maped types. Throws if invoked directly.
     */
    get ShallowEntityType(): ShallowEntityType<T> {
        throw getTypeAccessorError("ShallowEntityType", "MappedDataSource");
    }

    /**
     * Getter to extract static type of member entity
     *
     * This is only useful in maped types. Throws if invoked directly.
     */
    get EntityType(): NestedEntityType<T> {
        throw getTypeAccessorError("NestedEntityType", "MappedDataSource");
    }

    /**
     * Getter to extract static type of the mapping used to configure the data source
     *
     * This is only useful in maped types. Throws if invoked directly.
     */
    get MappingType(): T {
        throw getTypeAccessorError("MappingType", "MappedDataSource");
    }

    /**
     * Get the default GraphQL output type for a member entity
     */
    @MemoizeGetter
    get defaultOutputType(): GraphQLOutputType {
        return deriveDefaultOutputType(this);
    }

    /**
     * Get the output type for the response ofa paginated response performed against this data source
     */
    @MemoizeGetter
    get paginatedOutputType() {
        return derivePaginatedOutputType(this.pageContainerName, this.pageName, this.defaultOutputType);
    }

    /**
     * Get the output type for the response of a paginated query (for shallow entities) performed against this data source
     */
    @MemoizeGetter
    get paginatedShallowOutputType() {
        return derivePaginatedOutputType(
            this.shallowPageContainerName,
            this.shallowPageName,
            this.defaultShallowOutputType,
        );
    }

    /**
     * Get the default GraphQL input type for a shallow member entity (excludes associations)
     */
    @MemoizeGetter
    get defaultShallowInputType(): GraphQLInputType {
        debug("Deriving default shallow input type for: ", this.mappedName);
        return deriveDefaultShallowInputType(this);
    }

    /**
     * Get the default GraphQL output type for a shallow member entity (excludes associations)
     */
    @MemoizeGetter
    get defaultShallowOutputType(): GraphQLOutputType {
        return deriveDefaultShallowOutputType(this);
    }

    /**
     * Maps member entities (what the application interacts with) to rows (in the format the persistence layer expects)
     */
    mapEntitiesToRows(entities: ShallowEntityType<T>[]): Dict[] {
        return entities.map(entity =>
            reduce<ShallowEntityType<T>, Dict>(
                entity,
                (result: Dict, val, key: any) => {
                    const field = this.fields[key];
                    if (field) {
                        return field.reduce(result, val);
                    }
                    return result;
                },
                {},
            ),
        );
    }

    /**
     * Reverse map the rows obtained from the data source (the persistence layer) to member entities (what the application interacts with)
     */
    mapRowsToEntities(rows: Dict[], storeParams: StoreQueryParams<MappedDataSource<T>>) {
        return new ReverseMapper(storeParams).reverseMap(rows);
    }

    /**
     * Reverse map the rows obtained from the data source (the persistence layer) to shallow member entities
     * (what the application interacts with)
     *
     * This does not deal with mapping of associations, so is relatively cheaper than mapRowsToEntities
     */
    mapRowsToShallowEntities(rows: Dict[]) {
        return rows.map(row => {
            const mappedRow: Dict = {};
            for (const [, field] of Object.entries<MappedField<MappedDataSource<T>, FieldMapping<any, any>>>(
                this.fields,
            )) {
                mappedRow[field.mappedName] = field.getValue(row);
            }
            return mappedRow;
        });
    }

    /**
     * Given a query (entity field name -> value mapping), translates it into a query that the persistence
     * layer can process by mapping entity field names to source column names.
     */
    mapQueryParams(whereArgs: Dict, aliasHierarchyVisitor: AliasHierarchyVisitor) {
        const whereParams: Dict = {};
        Object.entries(whereArgs).forEach(([name, arg]) => {
            const field: MappedField = (this.fields as any)[name];
            if (field) {
                whereParams[`${aliasHierarchyVisitor.alias}.${field.sourceColumn}`] = arg;
                return;
            }
        });
        return whereParams;
    }
}

/**
 * Used to map a relational data source using specified configuration (of type [DataSourceMapping](api:DataSourceMapping)).
 *
 * Refer the guide on [Mapping Data Sources](guide:mapping-data-sources) for detailed explanation and examples.
 *
 * @api-category PrimaryAPI
 */
export const mapDataSource = <T extends DataSourceMapping>(mapping: T) => new MappedDataSource<T>(mapping);
