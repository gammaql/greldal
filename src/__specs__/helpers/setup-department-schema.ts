import * as Knex from "knex";
import { reportErrors } from "../test-helpers";

export const setupDepartmentSchema = async (knex: Knex) => {
    console.log("Setting up department schema: start");
    try {
        await knex.schema.createTable("tags", t => {
            t.increments("id");
            t.string("name");
        });
        await knex.schema.createTable("departments", t => {
            t.increments("id");
            t.string("name");
        });
        await knex.schema.createTable("products", t => {
            t.increments("id");
            t.integer("department_id")
                .unsigned()
                .references("id")
                .inTable("departments")
                .notNullable();
            t.string("name");
        });
        await knex.schema.createTable("product_tag_associators", t => {
            t.increments("id");
            t.integer("product_id")
                .unsigned()
                .references("id")
                .inTable("products")
                .notNullable();
            t.integer("tag_id")
                .unsigned()
                .references("id")
                .inTable("tags")
                .notNullable();
        });
    } catch (e) {
        console.error("Error when setting up departments schema");
        console.error(e);
        throw e;
    }
    console.log("Setting up department schema: end");
};

export const teardownDepartmentSchema = async (knex: Knex) => {
    await reportErrors(async () => {
        await knex.schema.dropTable("product_tag_associators");
        await knex.schema.dropTable("products");
        await knex.schema.dropTable("departments");
        await knex.schema.dropTable("tags");
    });
};
