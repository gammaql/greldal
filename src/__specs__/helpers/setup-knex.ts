import Knex from "knex";
import tmp from "tmp";
import assert from "assert";
import _debug from "debug";

import { Maybe } from "../../util-types";

const debug = _debug("greldal:setup-knex");

export const getConnectionString = () => {
    const connectionString = process.env.DB_CONNECTION_STRING;
    assert(connectionString, "Expected DB_CONNECTION_STRING environment variable to be defined");
    return connectionString;
};

export const setupKnex = () => {
    const isCI = process.env.CI === "true" || process.env.CI === "1";
    let config: Maybe<Knex.Config>;
    const db = process.env.DB;
    if (!db || db === "sqlite3") {
        const sqliteFile = tmp.fileSync();
        config = {
            client: "sqlite3",
            connection: {
                filename: sqliteFile.name,
                multipleStatements: true,
            },
            useNullAsDefault: true,
        };
    } else if (db === "mysql2") {
        if (isCI) {
            config = {
                client: "mysql2",
                connection: {
                    host: "127.0.0.1",
                    user: "travis",
                    database: "greldal_test",
                    password: "",
                },
            };
        } else {
            config = {
                client: "mysql2",
                connection: getConnectionString(),
            };
        }
    } else if (db === "pg") {
        if (isCI) {
            config = {
                client: "pg",
                connection: {
                    user: "postgres",
                    database: "greldal_test",
                },
            };
        } else {
            config = {
                client: "pg",
                connection: getConnectionString(),
            };
        }
    } else {
        throw new Error(`Invalid db selection: ${db}`);
    }
    debug("Connecting to knex using configuration: %O", config);
    assert(config, "Failed to configure database for the test suite. You may need to update above configuration");
    return Knex({
        ...config,
        debug: !!process.env.DEBUG,
        pool: {
            min: 0,
        },
    });
};
