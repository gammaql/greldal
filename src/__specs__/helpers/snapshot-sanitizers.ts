import { isString, isFunction } from "lodash";
import { isObject } from "util";
import Maybe from "graphql/tsutils/Maybe";

const replaceCode = (err: string) => err.replace(/\[GRelDAL:\S+Error:\d+\]/, "[GRelDAL:ERR_CODE_OMITTED]");

export const removeErrorCodes = (errors: Maybe<any[]>) =>
    (errors || []).map(err => {
        if (isString(err)) return replaceCode(err);
        if (isObject(err) && err.message) return replaceCode(err.message);
        if (isObject(err) && isFunction(err.toString)) return replaceCode(err.toString());
        return replaceCode(`${err}`);
    });
