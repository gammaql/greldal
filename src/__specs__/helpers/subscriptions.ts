import { ExecutionResult } from "graphql";
import { ExecutionResultDataDefault } from "graphql/execution/execute";
import { isFunction } from "lodash";

const isAsyncIterator = (i: any): i is AsyncIterator<any> => i && isFunction(i.next);

export const getSubscriptionResults = (count = 1) => async (
    resultIterator: ExecutionResult<ExecutionResultDataDefault> | AsyncIterator<ExecutionResult<ExecutionResultDataDefault>>
): Promise<ExecutionResult<ExecutionResultDataDefault>[]> => {
    if (!isAsyncIterator(resultIterator)) return [resultIterator];
    const returned: any = [];
    let idx = 0;
    while (true) {
        const { done, value } = await resultIterator.next();
        if (done) return returned;
        returned.push(value);
        idx += 1;
        if (idx === count) return returned;
    }
}