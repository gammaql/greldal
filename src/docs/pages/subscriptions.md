import {NextPageLink} from "../components/Link";
import Link from "next/link";
import {CodeSnippet} from "../components/CodeSnippet";

# Subscriptions

It is easy to use [graphql-subscriptions](https://github.com/apollographql/graphql-subscriptions) alongside GRelDAL to support real-time subscriptions.

<CodeSnippet name="mapSchema_insert_subscription" />

Currently there is no explicit support of tracking change in associated data sources. This is likely to be added in near future.

<NextPageLink>Best Practices</NextPageLink>
