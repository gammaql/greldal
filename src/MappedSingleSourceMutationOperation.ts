import { noop } from "lodash";

import { MappedSingleSourceOperation } from "./MappedSingleSourceOperation";
import { MappedDataSource } from "./MappedDataSource";
import { SourceAwareOperationResolver } from "./SourceAwareOperationResolver";
import { Dict } from "./util-types";
import { SourceAwareResolverContext } from "./SourceAwareResolverContext";
import { OperationType } from "./operation-types";

export enum MutationType {
    Insert = "INSERT",
    Update = "UPDATE",
    Delete = "DELETE"
}

export interface MutationPublishPayload {
    source: string;
    type: MutationType;
    primary: Dict[];
}

/**
 * @api-category MapperClass
 */
export abstract class MappedSingleSourceMutationOperation<
    TSrc extends MappedDataSource,
    TArgs extends {}
> extends MappedSingleSourceOperation<TSrc, TArgs> {
    constructor(
        public mapping: MappedSingleSourceOperation<TSrc, TArgs>["mapping"] & {
            resolver?: <
                TCtx extends SourceAwareResolverContext<MappedSingleSourceMutationOperation<TSrc, TArgs>, TSrc, TArgs>,
                TResolved
            >(
                ctx: TCtx,
            ) => SourceAwareOperationResolver<TCtx, TSrc, TArgs, TResolved>;
            publish?: (payload: MutationPublishPayload) => void;
        },
    ) {
        super(mapping);
    }
    operationType = OperationType.Mutation;

    get publish() {
        return this.mapping.publish || noop;
    }
}
