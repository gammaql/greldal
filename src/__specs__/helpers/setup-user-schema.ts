import * as Knex from "knex";
import * as types from "../../types";
import { mapDataSource } from "../../MappedDataSource";
import { GraphQLID, GraphQLString, GraphQLInt } from "graphql";
import { times } from "lodash";
import { mapFields } from "../..";

export const setupUserSchema = async (knex: Knex) => {
    await knex.schema.createTable("users", t => {
        t.increments("id");
        t.string("name");
        t.integer("age");
        t.jsonb("metadata");
    });
};

export const insertFewUsers = async (knex: Knex) => {
    await knex("users").insert([
        {
            id: 1,
            name: "Lorefnon",
            age: 50,
            metadata: JSON.stringify({
                positionsHeld: [
                    {
                        title: "Software Architect",
                        organization: "Foo Bar Inc",
                        duration: 5,
                    },
                    {
                        title: "Software Developer",
                        organization: "Lorem Ipsum Gmbh",
                        duration: 10,
                    },
                ],
                awards: [
                    {
                        title: "Top Achiever",
                        compensation: 1000,
                    },
                ],
            }),
        },
        { id: 2, name: "Gandalf", age: 1000 },
    ]);
};

export const insertManyUsers = async (knex: Knex) => {
    let users = [];
    for (const id of times(500)) {
        users.push({
            id,
            name: `User ${id}`,
            age: 50,
            metadata: JSON.stringify({
                positionsHeld: times(5).map(idx => ({
                    title: `Pos ${id}:${idx}`,
                    organization: `Org ${id}:${idx}`,
                    duration: 3,
                })),
                awards:
                    id > 100
                        ? []
                        : times(3).map(idx => ({
                              title: `Award: ${idx}`,
                              compensation: 100,
                          })),
            }),
        });
        if (users.length >= 20) {
            await knex("users").insert(users);
            users = [];
        }
    }
};

export const mapUsersDataSource = () => {
    // @snippet:start mapDataSource_user_simple
    /// import {mapDataSource, mapFields, types} from "greldal";

    const users = mapDataSource({
        name: "User",
        fields: mapFields({
            id: {
                type: types.number,
                to: GraphQLID,
                isPrimary: true,
            },
            name: {
                type: types.string,
            },
            age: {
                type: types.integer,
            },
        }),
    });
    // @snippet:end
    return users;
};

export const mapUsersDataSourceExplicitly = () => {
    // @snippet:start mapDataSource_user_simple_explicit
    /// import {mapDataSource, mapFields, types} from "greldal";

    const users = mapDataSource({
        name: {
            mapped: "User",
            stored: "users",
        },
        fields: mapFields({
            id: {
                sourceColumn: "id",
                type: types.string,
                to: {
                    input: GraphQLID,
                    output: GraphQLID,
                },
            },
            name: {
                sourceColumn: "name",
                type: types.string,
                to: {
                    input: GraphQLString,
                    output: GraphQLString,
                },
            },
            age: {
                sourceColumn: "age",
                type: types.integer,
                to: {
                    input: GraphQLInt,
                    output: GraphQLInt
                }
            }
        }),
    });
    // @snippet:end
    return users;
};

export const mapUsersDataSourceWithJSONFields = () =>
    mapDataSource({
        name: "User",
        fields: mapFields({
            id: {
                type: types.number,
                to: GraphQLID,
                isPrimary: true,
            },
            name: {
                type: types.string,
            },
            age: {
                type: types.integer,
            },
            metadata: {
                type: types.maybe(
                    types.json(
                        types.partial({
                            positionsHeld: types.array(
                                types.interface({
                                    title: types.string,
                                    organization: types.string,
                                    duration: types.integer,
                                }),
                            ),
                            awards: types.array(
                                types.interface({
                                    compensation: types.number,
                                    title: types.string,
                                }),
                            ),
                        }),
                    ),
                ),
            },
        }),
    });

export const teardownUserSchema = async (knex: Knex) => {
    await knex.schema.dropTable("users");
};
