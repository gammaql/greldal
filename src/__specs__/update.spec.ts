import { GraphQLSchema, GraphQLString, GraphQLList, subscribe, parse, graphql, GraphQLObjectType, GraphQLInt, GraphQLID } from "graphql";
import Knex from "knex";
import { PubSub } from 'graphql-subscriptions'

import { MappedDataSource } from "../MappedDataSource";
import { setupUserSchema, insertFewUsers, mapUsersDataSource, teardownUserSchema } from "./helpers/setup-user-schema";
import { mapSchema, operationPresets, useDatabaseConnector } from "..";
import { setupKnex } from "./helpers/setup-knex";
import { getSubscriptionResults } from './helpers/subscriptions';
import { MutationPublishPayload } from '../MappedSingleSourceMutationOperation';

let knex: Knex;

jest.setTimeout(30000);

describe("Update operation", () => {
    let users: MappedDataSource, schema: GraphQLSchema;
    const pubsub = new PubSub()

    beforeAll(() => {
        knex = setupKnex();
        useDatabaseConnector(knex);
    });

    describe("Subscriptions", () => {
        beforeAll(async () => {
            await setupUserSchema(knex);
            await insertFewUsers(knex);
            users = mapUsersDataSource();
            schema = mapSchema([
                operationPresets.findOneOperation(users),
                operationPresets.updateOneOperation(users, (mapping) => ({
                    ...mapping,
                    publish: (payload: MutationPublishPayload) => pubsub.publish("MUTATIONS", payload)
                })),
                {
                    operationType: "subscription" as const,
                    name: "userUpdated",
                    fieldConfig: {
                        type: GraphQLList(new GraphQLObjectType({
                            name: "UserUpdateNotification",
                            fields: {
                                id: {
                                    type: GraphQLID
                                }    
                            }
                        })),
                        resolve: (payload) => {
                            return payload.primary;
                        },
                        subscribe: () => pubsub.asyncIterator("MUTATIONS")
                    }
                }
            ]);
        });
        afterAll(async () => {
            await teardownUserSchema(knex);
        });
        test("Updates are published as a mutation via configured publisher", async () => {
            const subscriptionQuery = `
                subscription {
                    userUpdated {
                        id
                    }
                }
            `;
            const subP = subscribe(schema, parse(subscriptionQuery)).then(getSubscriptionResults());
            const graphQLResult = await graphql(schema, `
                mutation {
                    updateOneUser(where: {id: 1}, update: { 
                        name: "Sherlock Holmes" 
                    }) {
                        id,
                        name    
                    }
                }
            `);
            const subscriptionResult = await subP;
            expect(graphQLResult.data!.updateOneUser.id).toEqual(subscriptionResult[0].data!.userUpdated[0].id);
            expect(graphQLResult).toMatchSnapshot();
            expect(subscriptionResult).toMatchSnapshot();
        });
    });
});