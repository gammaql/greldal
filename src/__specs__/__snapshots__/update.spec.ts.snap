// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update operation Subscriptions Updates are published as a mutation via configured publisher 1`] = `
Object {
  "data": Object {
    "updateOneUser": Object {
      "id": "1",
      "name": "Sherlock Holmes",
    },
  },
}
`;

exports[`Update operation Subscriptions Updates are published as a mutation via configured publisher 2`] = `
Array [
  Object {
    "data": Object {
      "userUpdated": Array [
        Object {
          "id": "1",
        },
      ],
    },
  },
]
`;
